src/attributes.jl

```diff
@@ -1856,6 +1856,88 @@ function attribute_value_type(::ConstraintConflictStatus)
     return ConflictParticipationStatusCode
 end
 
+"""
+    UserDefinedFunction(name::Symbol, arity::Int) <: AbstractModelAttribute
+
+Set a user-defined function by the name of `name` with `arity` arguments.
+
+The value to be set is a tuple containing one, two, or three functions.
```

Comment:
For the case of a symbolic function recognized by a particular solver and for which we don't want to bother specifying evaluation (and maybe can't give derivatives), can we make things convenient at the MOI level? I'm looking [here in the Minizinc logical operators PR](https://github.com/jump-dev/MiniZinc.jl/pull/22/files#diff-3b9314a6f9f2d7eec1d0ef69fa76cfabafdbe6d0df923768f9ec32f27a249c63R1143) and thinking it would be cleaner to be able to specify just `MOI.Nonlinear.register_operator(nlp, :!, 1)`.

Reply:
Maybe I'm confused here by the name and docstring... is the purpose of UserDefinedFunction simply to allow passing in eval+derivs for solvers that require these? i.e. is it optional to specify UserDefinedFunction, and not needed for solvers that recognize function symbols directly (like MiniZinc)? If so, then I feel like UserDefinedFunction should be renamed to something like UserDefinedFunctionEvaluators to be a lot clearer (or the interface for specifying arity, f, ∇f, ∇²f should be redesigned), and the docstring `Set a user-defined function by the name of `name` with `arity` arguments.` should be rewritten to make it clear when/why to set these.

Reply:
Thinking on this more, I think the function evaluation is always required. Otherwise, how is the solver meant to know what function `!` refers to? With the introduction of `ListOfSupportedNonlinearOperators`, solvers can pre-define a list of operators they support and recognize.